connect/proxy/listener_test.go

```diff
@@ -19,71 +19,99 @@ import (
+	buf := &bytes.Buffer{}
+	replacer := strings.NewReplacer(" ", "_")
+
+	if len(parts) > 0 {
+		replacer.WriteString(buf, parts[0])
+	}
+	for _, part := range parts[1:] {
+		replacer.WriteString(buf, ".")
+		replacer.WriteString(buf, part)
+	}
+
+	key := buf.String()
+
+	for _, label := range labels {
+		replacer.WriteString(buf, fmt.Sprintf(";%s=%s", label.Name, label.Value))
+	}
+
+	return buf.String(), key
+}
+
+func testSetupMetrics(t *testing.T) *MockSink {
 	// Record for ages (5 mins) so we can be confident that our assertions won't
 	// fail on silly long test runs due to dropped data.
-	s := metrics.NewInmemSink(10*time.Second, 300*time.Second)
+	s := MockSink{}
 	cfg := metrics.DefaultConfig("consul.proxy.test")
 	cfg.EnableHostname = false
-	metrics.NewGlobal(cfg, s)
-	return s
+	metrics.NewGlobal(cfg, &s)
+	return &s
 }
 
-func assertCurrentGaugeValue(t *testing.T, sink *metrics.InmemSink,
+func assertTelemetryValue(t *testing.T, sink *MockSink,
 	name string, value float32) {
 	t.Helper()
 
-	data := sink.Data()
+	var idx int
+	var found bool
 
-	// Current interval is the last one
-	currentInterval := data[len(data)-1]
-	currentInterval.RLock()
-	defer currentInterval.RUnlock()
-
-	assert.Equalf(t, value, currentInterval.Gauges[name].Value,
-		"gauge value mismatch. Current Interval:\n%v", currentInterval)
-}
+	for i := 0; i < len(sink.keys); i++ {
+		k, _ := sink.flattenKeyLabels(sink.keys[i], sink.labels[i])
 
-func assertAllTimeCounterValue(t *testing.T, sink *metrics.InmemSink,
-	name string, value float64) {
-	t.Helper()
-
-	data := sink.Data()
-
-	var got float64
-	for _, intv := range data {
-		intv.RLock()
-		// Note that InMemSink uses SampledValue and treats the _Sum_ not the Count
-		// as the entire value.
-		if sample, ok := intv.Counters[name]; ok {
-			got += sample.Sum
+		if k == name {
+			idx = i
+			found = true
 		}
-		intv.RUnlock()
 	}
 
-	if !assert.Equal(t, value, got) {
-		// no nice way to dump this - this is copied from private method in
-		// InMemSink used for dumping to stdout on SIGUSR1.
-		buf := bytes.NewBuffer(nil)
-		for _, intv := range data {
-			intv.RLock()
-			for _, val := range intv.Gauges {
-				fmt.Fprintf(buf, "[%v][G] '%s': %0.3f\n", intv.Interval, name, val.Value)
-			}
-			for name, vals := range intv.Points {
-				for _, val := range vals {
-					fmt.Fprintf(buf, "[%v][P] '%s': %0.3f\n", intv.Interval, name, val)
-				}
-			}
-			for _, agg := range intv.Counters {
-				fmt.Fprintf(buf, "[%v][C] '%s': %s\n", intv.Interval, name, agg.AggregateSample)
-			}
-			for _, agg := range intv.Samples {
-				fmt.Fprintf(buf, "[%v][S] '%s': %s\n", intv.Interval, name, agg.AggregateSample)
-			}
-			intv.RUnlock()
+	dumpSink := func() {
+		t.Log("Sink contents:")
+		for i := 0; i < len(sink.keys); i++ {
+			k, _ := sink.flattenKeyLabels(sink.keys[i], sink.labels[i])
+			t.Logf("\t%s: %f", k, sink.vals[i])
 		}
-		t.Log(buf.String())
+	}
```

Comment:
Why not make this a free function (or ideally a method on `MockSink` here or in `go-metrics` in the eventual PR). So it can be re-used?